abcpy/ProbabilisticModel.py

```diff
@@ -0,0 +1,719 @@
+from abc import ABCMeta, abstractmethod
+import numpy as np
+
+from scipy.stats import multivariate_normal, norm
+from scipy.special import gamma
+
+#TODO both ricker and lorenz could support that you give some in a combined model, and some not, we should implement that, but not a priority
+
+#TODO docstrings
+
+#NOTE do we maybe want to average over a couple of samples during initialization, rather than taking a single value? this is an issue for example for StudentT!
+
+class ProbabilisticModel(metaclass = ABCMeta):
+    def __init__(self, parameters):
+        self.parents = parameters #all parents of the node
+        self.updated = False #whether the node has been touched
+
+    #NOTE the check_fixed assumes that the parameters are exactly of the lenght needed for that distribution only ->needs to be considered in case we want to be able to "fix on whole graph"
+    def fix_parameters(self, parameters=None, rng=np.random.RandomState()):
+        if(not(parameters)):
+            self.value = self.sample_from_distribution(1, rng)
+            if(isinstance(self.value[0],np.ndarray)):
+                self.value = self.value[0]
+        else:
+            self.check_parameters_fixed(parameters)
+            index=0
+            i=0
+            while(i<len(parameters)):
```

Comment:
As far as I can see you store the fixed parameter of a model in its parents models/ parameters. I would find it more natural to get a fixed value from the parent and store it as self.parameters_fixed or so. What do you think?

Reply:
I store the "fixed" output of each node/probabilistic model as the value of that node. To me, that seems more natural than saving information about the parents, since it is the value that is directly associated with the node, while the information about the parents comes from the parents. But I can see that it might make sense to have the parameter values as an attribute.

Since we have to check all parents for their type anyways in most circumstances (setting them and getting them, not for simulation), it won't make a huge difference in amount of calls.

However, if we store all parameter values as a attribute as well, we kind of have information twice: we need to know the parents, so we need to store those. If we have n parents, we will then also store n values for each parent, so more data gets sent around.

I think we discussed this last week, and agreed to do it as I am doing it now, but I can easily change it if you wish.

Reply:
Good. We have to store the fixed values anyway, so I am not seeing why we should have twice the storage footprint if we store it at the model instead of the parent?

Reply:
Sorry, too early in the morning, we will store the same amount of data in total. Will just be distributed a bit differently.